### PR TITLE
feat(rust): add accumulated context governance

### PR DESCRIPTION
**Pull request by lxingy3**
## Description  Adds the remaining Rust SDK slice of accumulated-context governance tracked by #3084.  - reuses the existing `DataClassification` type, completing its ordered ladder with `TopSecret` - adds an immutable, versioned `ContextEnvelope` with max-lattice sensitivity and grow-only restrictions - evaluates organization-authored label combinations and escalates unknown combinations at a configurable threshold - folds actual post-execution result labels, then gates later exports, delegations, and memory writes - inherits parent restrictions across delegation boundaries - gates actions and builds obligations from the rule-augmented restriction set, then maps constrained outcomes onto the existing Rust `PolicyDecision` surface without failing open - exposes an opaque `EnvelopeReference` containing only the envelope ID and aggregate sensitivity - emits classified context-transition audit events with exact cross-SDK wire values and delta-only payloads  The current Rust policy surface does not carry `result_labels`, despite the older issue text. This API therefore accepts actual labels explicitly after execution rather than guessing an action's output labels in advance.  Labels remain deterministic organization-authored category tokens backed by `BTreeSet<String>`. This avoids adding an unused parallel `DataLabel`/ABAC surface. No new dependencies are introduced.  Compatibility note: adding `TopSecret` extends a public Rust enum, so downstream exhaustive matches must add an arm. Existing discriminants do not change.  ## Type of Change - [ ] Bug fix (non-breaking change that fixes an issue) - [ ] New feature (non-breaking change that adds functionality) - [x] Breaking change (fix or feature that would cause existing functionality to change) - [x] Documentation update - [ ] Maintenance (dependency updates, CI/CD, refactoring) - [ ] Security fix  ## Package(s) Affected - [x] agent-mesh (`agent-governance-rust/agentmesh`) - [ ] agent-os-kernel - [ ] agent-runtime - [ ] agent-sre - [ ] agent-governance - [ ] docs / root  ## Verification  - `cargo test --release --workspace` — 524 passed, 0 failed - `cargo build --release --workspace` - `cargo clippy --release -p agentmesh --lib --no-deps -- -A deprecated` — no new warnings - `rustfmt --edition 2021 --check agentmesh/src/context.rs agentmesh/src/context_audit.rs` - `git diff --check`  ## Checklist - [x] My code follows the project style guidelines (Rustfmt and Clippy; Ruff is not applicable) - [x] I have added tests that prove my feature works - [x] All new and existing tests pass - [x] I have updated documentation as needed - [x] I have signed the [Microsoft CLA](https://cla.opensource.microsoft.com/)  ## Attribution & Prior Art - [x] This contribution does not contain code copied or derived from other projects without attribution - [x] Any external projects that inspired this design are credited in code comments or documentation - [x] If this PR implements functionality similar to an existing open-source project, I have listed it below  **Prior art / related projects:**  This is a Rust semantic-parity implementation of the in-repository Python context and audit design merged in #2800. The module documentation credits that source directly. No external project code was used.  ## AI Assistance - [x] I can explain every meaningful change in this PR: what it does, why, and what tradeoffs were considered - [x] I have run tests and verification appropriate for this change - [x] No part of this PR was autonomously submitted by an AI agent without my review - [x] I have not used AI to generate review comments on others' PRs  Codex assisted with implementation and test drafting. I reviewed the complete diff and validation results before submission.  ## IP, Patents, and Licensing - [x] This contribution does not implement patent-pending or patent-encumbered techniques - [x] This contribution does not require an NDA or licensing agreement to understand or use - [x] Any AI tools used have terms compatible with the MIT License  ## Related Issues  Part of #3084.

### CODE REVIEW

File: agent-governance-rust/agentmesh/src/context.rs [hunk 1]
```diff
@@ -0,0 +1,565 @@
...
+        _ => None,
+    };
+    let restriction_present =
+        gating_restriction.is_some_and(|restriction| envelope.restrictions.contains(restriction));
+    let floor_triggered =
+        gating_restriction.is_some() && aggregation.aggregate_sensitivity >= restricted_floor;
+
+    if restriction_present || floor_triggered {
+        let reason = if restriction_present {
+            format!(
```
Comment (MohammadHaroonAbuomar) [inline agent-governance-rust/agentmesh/src/context.rs:309]: nit: the constrain reason renders double quotes via {:?} vs Python's !r single quotes; the escalation reason at line 282 is byte-identical. Only matters if cross-SDK log matching ever does.

**Review by MohammadHaroonAbuomar — CHANGES_REQUESTED** [review]
- agent-governance-rust/agentmesh/src/context.rs: the context_audit surface is missing entirely. The Python reference ships context_audit.py (six CONTEXT_*/DERIVED_ARTIFACT_LABELED wire-string constants, a ContextEvent record with previous/new sensitivity, labels_added, rules_applied, restrictions_added, and a context_event(before, after) builder with classification = max(before, after)), and all three sibling ports include it (TS #3199 context-audit.ts, Go #3261 context_audit.go, .NET #3443 ContextEvent.cs). Add the audit module with the exact Python wire strings plus tests.

**Comment by lxingy3** [comment]
The concrete restriction-handling issue is fixed in 4ad6691: decide_next now gates and builds obligations from the rule-augmented restriction set, with a regression test for the case where those restrictions have not yet been persisted into the envelope. The requested audit surface is in 7279e3a. The full Rust workspace now passes 524 tests. I have left escalation mapping, token normalization, threshold edge cases, and classification wire representation unchanged pending the maintainer decision in #3084.

File: agent-governance-rust/agentmesh/src/context.rs [hunk 2]
```diff
@@ -0,0 +1,565 @@
...
+            aggregate_sensitivity = aggregate_sensitivity.max(rule.sets_sensitivity);
+            restrictions.extend(rule.adds_restrictions.iter().cloned());
+            rules_applied.push(rule.name.clone());
+        }
+    }
+
+    AggregationResult {
+        aggregate_sensitivity,
+        restrictions,
+        escalate: rules_applied.is_empty() && envelope.labels.len() >= category_threshold,
```
[the remark this answers is not in the document]
Comment (lxingy3) [reply agent-governance-rust/agentmesh/src/context.rs]: Leaving this open pending the cross-SDK threshold decision in #3084. The current Rust behavior matches the Python, TypeScript, and Go ports described there, so changing it here now would create a different parity mismatch.

**Comment by liamcrumm** [comment]
Two findings:

1. `AggregationRule` doesn't reject an empty `all_labels`, and `evaluate_aggregation` matches with `all_labels.is_subset(&envelope.labels)`. An empty set is a subset of everything, so such a rule fires on every envelope, `rules_applied` is never empty, and the unknown-combination escalation is silently disabled. #3524 adds exactly this guard on the Python side.
2. `escalate: rules_applied.is_empty() && envelope.labels.len() >= category_threshold` is always true when `category_threshold == 0`, so an empty envelope escalates.

`to_policy_decision` correctly guards the empty-obligation case, matching the .NET version.

**Comment by lxingy3** [comment]
Thanks, both are fixed in e3f05fe. `AggregationRule::new` now rejects an empty label set, with private fields so construction cannot bypass validation, and an empty envelope no longer escalates when the threshold is zero. I also rebased onto current `main`; `cargo test --workspace --locked` passes all 528 tests.

**Comment by Ricky-G** [comment]
This PR has been stale for more than a month (last updated July 31, 2026), so I’m closing it in favor of the fresh replacement PR #4065.

#4065 carries the complete Rust accumulated-context implementation on current `main`, is authored under my account, and includes the final fail-closed regression fix identified during review. I’ll keep the replacement actively tracked through CI, review, and merge so this Rust slice reaches the finish line cleanly.

The umbrella issue #3084 remains the tracker because the TypeScript slice is still outstanding.
